WD-Firmware/comrv/comrv.c

```diff
@@ -18,34 +18,767 @@
+   u32_t               uiProfilingIndication;
+#endif /* D_COMRV_FW_INSTRUMENTATION */
+#ifdef D_COMRV_MULTI_GROUP_SUPPORT
+   u32_t               uiTemp;
+   u16_t               usSelectedMultiGroupEntry;
+#endif /* D_COMRV_MULTI_GROUP_SUPPORT */
+
+   /* read the requested token value (t5) */
+   M_COMRV_READ_TOKEN_REG(unToken.uiValue);
+
+   /* read comrv stack register (t3) */
+   M_COMRV_READ_STACK_REG(pComrvStackFrame)
+
+   /* get the invoke callee indication */
+   ucIsInvoke = (u32_t)pComrvStackFrame & D_COMRV_INVOKE_CALLEE_BIT_0;
+   /* are we calling a new overlay function */
+   if (ucIsInvoke == D_COMRV_INVOKE_CALLEE_BIT_0)
+   {
+      /* clear the invoke indication from pComrvStackFrame address */
+      pComrvStackFrame = (comrvStackFrame_t*)((u32_t)pComrvStackFrame & (~D_COMRV_PROFILING_INVOKE_VAL));
+      /* write back the stack register after bit 0 was cleared*/
+      M_COMRV_WRITE_STACK_REG(pComrvStackFrame);
+
+#ifdef D_COMRV_FW_INSTRUMENTATION
+      uiProfilingIndication = D_COMRV_NO_LOAD_AND_INVOKE_IND;
+#endif /* D_COMRV_FW_INSTRUMENTATION */
+   }
+   /* we are returning to an overlay function or accessing overlay data */
+   else
+   {
+#ifdef D_COMRV_FW_INSTRUMENTATION
+      uiProfilingIndication = D_COMRV_NO_LOAD_AND_RETURN_IND;
+#endif /* D_COMRV_FW_INSTRUMENTATION */
+   }
+
+#ifdef D_COMRV_MULTI_GROUP_SUPPORT
+   /* if the requested token isn't a multi-group token */
+   if (unToken.stFields.multiGroup == 0)
+   {
+#endif /* D_COMRV_MULTI_GROUP_SUPPORT */
+      /* search for token */
+      pAddress = comrvSearchForLoadedOverlayGroup(unToken);
+#ifdef D_COMRV_MULTI_GROUP_SUPPORT
+   }
+   /* search for a multi-group overlay token */
+   else
+   {
+      /* first ucEntryIndex to search from in the multi group table is determined by the overlayGroupID
+         field of the requested token */
+      ucEntryIndex = unToken.stFields.overlayGroupID;
+      do
+      {
+         /* search for the token */
+         pAddress = comrvSearchForLoadedOverlayGroup(pOverlayMultiGroupTokensTable[ucEntryIndex++]);
+      /* continue the search as long as the group wasn't found and we have additional tokens */
+      } while ((pAddress == NULL) && (pOverlayMultiGroupTokensTable[ucEntryIndex].uiValue != 0));
+
+      /* save the selected multi group entry */
+      usSelectedMultiGroupEntry = ucEntryIndex-1;
+   }
+#endif /* D_COMRV_MULTI_GROUP_SUPPORT */
+
+   /* get the group size */
+   usOverlayGroupSize = M_COMRV_GET_OVL_GROUP_SIZE(unToken);
+
+   /* if the data/function is not loaded we need to evict and load it */
+   if (pAddress == NULL)
+   {
+#ifdef D_COMRV_MULTI_GROUP_SUPPORT
+      /* if the requested token is a multi-group token */
+      if (unToken.stFields.multiGroup != 0)
+      {
+         /* for now we take the first token in the list of tokens */
+         unToken = pOverlayMultiGroupTokensTable[unToken.stFields.overlayGroupID];
+         /* save the selected multi group entry */
```

Comment:
explain why we save the entry in the comment

Reply:
done in 14d152f